packages/core/src/new-api/internal/execution/execution-engine.ts

```diff
@@ -262,18 +290,23 @@ export class ExecutionEngine {
       await this._apply(state, initMessage);
     }
 
-    while (!this._isBatchComplete(batch, state)) {
+    while (!this._isBatchComplete(state, batch)) {
       const sortedFutures: Future[] = sortFuturesByNonces(batch, state);
 
       const results = await this._submitOrCheckFutures(sortedFutures, state);
 
       batchResults = [...batchResults, ...results];
 
-      if (this._isBatchComplete(sortedFutures, state)) {
+      if (this._isBatchComplete(state, sortedFutures)) {
         break;
       }
 
-      await this._newBlock(state);
+      const timeoutResults = await this._newBlockOrTimeout(
```

Comment:
This needed enhanced. If the timeout comes between blocks, we need check for timeouts in this wait period.

This also implies it must return timeout result messages so they can be gathered together for the batch.

Reply:
A refactor we should consider is not passing result messages around here, and calculate the final result based on the execution state. Actually let me add a todo.

Reply:
Yes, I think that would be good.

Reply:
I have captured a cleanup task here: https://github.com/NomicFoundation/ignition/issues/320